Pills do not show selected value(s) if disabled #12388

davidsjoberg1 opened the pull request:
## Describe your changes
Before the change, when disabling a st.pills component it turned into a faded gray color. The goal with the issue was to make still make it faded but keep the same colors as when not disabled. 

In styled-components.ts for BaseButton there is a style called StyledButtonGroupBaseButton where the colors was set when a button is disabled. Instead of changing color I when disabling a button I just used opacity to make the faded look. Since StyledButtonGroupBaseButton is used in both st.pills and st.segmented_control this change will affect both of these components. 

### disabled=False
This is how it looks when the button is not disabled (just like before)

https://github.com/user-attachments/assets/75bb71e9-2493-4c3b-bb1f-6f61ebc2c7ff

### disabled=True 
Before change:

https://github.com/user-attachments/assets/5f5e7d32-4d1a-4b8a-869a-d951e1fec00f 

After change:

https://github.com/user-attachments/assets/9d039b62-d5ea-4560-8547-127f7dca25b6



<!-- If it's a visual change, please include a screenshot or video! -->

## GitHub Issue Link 
Closes https://github.com/streamlit/streamlit/issues/12388

## Testing Plan

I have made changes in e2e_playwright/st_pills_test.py and e2e_playwright/st_segmented_control_test.py. 
I updated the disabled tests to expect opacity instead of negating an rgb(...) color, matching the new design. 

- Unit Tests (JS and Python) ✅ 
- E2E Tests: st_segmented_control_test.py ✅ 
- E2E Tests: st_pills_test.py ✅ 


---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

davidsjoberg1 committed:
in StyledButtonGroupBaseButton changed from color to opacity when disabled

davidsjoberg1 committed:
changed opacity to 0.4. changed pills and segmented control tests to check opacity instead of color

davidsjoberg1 committed:
Merge branch 'develop' into gh-12388

jrieke commented:
Hey @davidsjoberg1! Good catch, thanks for the contribution! I agree we should change this, but I think we need a different approach than simply adding opacity, so this is more consistent with other widgets. 

Importantly, the disabled pills/controls should 1) not have a hover effect and 2) be completely grayscale. (That's also what we do for all other widgets). 

Maybe it would already be sufficient if you use a slightly gray background for the selected pills/controls instead of the current white?

davidsjoberg1 committed:
Reset the previous changes in the e2e tests. Modified the disabled StyledButtonGroupBaseButton to match disabled StyledPrimaryButton. Added lightGray background on active pills/segmented controls

davidsjoberg1 committed:
Merge branch 'develop' into gh-12388

davidsjoberg1 committed:
merge

davidsjoberg1 committed:
Merge branch 'gh-12388' of github.com:davidsjoberg1/streamlit into gh-12388

davidsjoberg1 committed:
matched active pills/segment control to header button

davidsjoberg1 committed:
reverted changes from last commit

davidsjoberg1 commented:
Hello @jrieke, thank you for the feedback!
I have now made some changes:

1. The color for disabled pills/control now matches StyledPrimaryButton, to keep things consistent
2. When a disabled pill/control is selected, the background color is gray30. Please let me know if you prefer another color!
3. There is no hover effect
4. The previous changes in the test files are reset

Below is a video of how the current implementation looks

https://github.com/user-attachments/assets/2ca86d7f-93dc-4122-b905-301d056c2075



Please let me know if I need to make some changes, or if I missunderstood something :)

jrieke commented:
Looks great! I think we should use a different color than `gray30` to make this theme-aware. It seems like the tags in `st.multiselect` use `theme.colors.fadedText05` for the background when disabled, I just tried this out locally and it looks pretty good, wdyt?

<img width="367" height="235" alt="Screenshot 2025-09-22 at 20 19 34" src="https://github.com/user-attachments/assets/7cf5b798-7e22-4c9b-bb13-9362542f109b" />

davidsjoberg1 committed:
changed active disabled pills/control from gray30 to fadedText05

davidsjoberg1 commented:
I agree! I just pushed the change to fadedText05

davidsjoberg1 committed:
Merge branch 'develop' into gh-12388

sfc-gh-bnisco commented:
Thanks for the contribution @davidsjoberg1!

Can you please add Snapshot tests for this case? I see that we already have a test case for disabled `st.pills`, but not for disabled `st.pills` with a selected value.

We have directions in our contributing wiki: https://github.com/streamlit/streamlit/wiki/Running-e2e-tests-and-updating-snapshots. Let me know if you have questions, thank you!

davidsjoberg1 committed:
added test for selected and disabled pill

davidsjoberg1 committed:
Merge branch 'develop' into gh-12388

davidsjoberg1 committed:
added snapshots for disabled-selected. updated snapshot for disabled

davidsjoberg1 committed:
moved disabled-selected pill test further up so it comes after disabled test

davidsjoberg1 commented:
Hello @sfc-gh-bnisco! 
I just added a test for disabled-selected pills. I basically copied the test for disabled but with the disabled-selected ```st.pills```. I added the disabled-selected ```st.pills``` in the end of ```st_pills.py```. It would make sense to put this after the disabled ```st.pills```, however then I need to change all indexes in ```st_pills_test.py```. Do you want me to make this change or is it ok with the current one?

sfc-gh-bnisco committed:
Update snapshots

sfc-gh-bnisco commented:
Thanks for checking @davidsjoberg1 , that's a really thoughtful question! The way you handled it is nice and pragmatic!

Long-term we are moving our tests away from the index-based paradigm anyway, but this test suite hasn't yet been updated. We are good to leave this as-is!

davidsjoberg1 committed:
Merge branch 'develop' into gh-12388

jrieke commented:
@davidsjoberg1 Congrats on getting this merged! 🥳 We'd love to send you some swag as a little thank you, just fill out this form and we'll get it on the way: https://docs.google.com/forms/d/e/1FAIpQLSe7cXh3H1DmrgNpVew9qViGIHX7vdEbTv5isA44_z5bgaKTKg/viewform

davidsjoberg1 commented:
@jrieke Nice, thank you!
